{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**MMSEQS2**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mmseqs/bin/mmseqs easy-search impairment_of_the_mitochondrial_integrity_and_biogenesis_genes.fasta realmergedprots_noen.faa mitimpair_noen.m8 -c 0.2 tmp --min-seq-id 0.3 --alt-ali 25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     qseqid            sseqid  pident  length  mismatch  \\\n",
      "0  59463.ENSMLUP00000016363  mlXP_023610807.1   0.961     389        15   \n",
      "1  59463.ENSMLUP00000016363  mmXP_036164646.1   0.953     387        18   \n",
      "2  59463.ENSMLUP00000016363  mdXP_059550832.1   0.948     387        20   \n",
      "3  59463.ENSMLUP00000016363  pkXP_036265305.1   0.935     387        25   \n",
      "4  59463.ENSMLUP00000016363  moXP_036135537.1   0.912     387        34   \n",
      "\n",
      "   gapopen  qstart  qend  sstart  send         evalue  bitscore  \n",
      "0        0       1   387       1   389  3.536000e-254       780  \n",
      "1        0       1   387       1   387  1.551000e-252       775  \n",
      "2        0       1   387       1   387  9.320000e-251       770  \n",
      "3        0       1   387       1   386  1.339000e-244       752  \n",
      "4        0       1   387       1   387  1.546000e-239       738  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "columns = [\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \n",
    "           \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"]\n",
    "\n",
    "df = pd.read_csv('batstring.m8', sep='\\t', names=columns)\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       qseqid            sseqid  pident  length  mismatch  \\\n",
      "0    59463.ENSMLUP00000000083  rfXP_032946134.1   0.964     284        10   \n",
      "1    59463.ENSMLUP00000001427  rfXP_032943217.1   0.931     189        13   \n",
      "2    59463.ENSMLUP00000001789  rfXP_032971332.1   0.959     524        21   \n",
      "3    59463.ENSMLUP00000002616  rfXP_032990906.1   0.929     427        30   \n",
      "4    59463.ENSMLUP00000002847  rfXP_032969289.1   0.807     209        39   \n",
      "..                        ...               ...     ...     ...       ...   \n",
      "488  59463.ENSMLUP00000014213    pnCAK6441801.1   0.995     660         3   \n",
      "489  59463.ENSMLUP00000016071    pnCAK6440610.1   0.835     462        75   \n",
      "490  59463.ENSMLUP00000016363    pnCAK6444154.1   0.975     248         6   \n",
      "491  59463.ENSMLUP00000016529    pnCAK6437761.1   0.912     663        58   \n",
      "492  59463.ENSMLUP00000022559    pnCAK6446407.1   0.980     157         3   \n",
      "\n",
      "     gapopen  qstart  qend  sstart  send         evalue  bitscore  \n",
      "0          0       1   284      94   377  1.343000e-183       571  \n",
      "1          0      24   212       1   189  1.609000e-118       379  \n",
      "2          0       1   524       1   524   0.000000e+00      1043  \n",
      "3          0       1   427       1   427  4.627000e-270       828  \n",
      "4          0       1   209       1   203  3.690000e-105       340  \n",
      "..       ...     ...   ...     ...   ...            ...       ...  \n",
      "488        0       1   660      21   680   0.000000e+00      1330  \n",
      "489        0       1   462       1   453  3.428000e-241       747  \n",
      "490        0     140   387       1   248  5.966000e-157       499  \n",
      "491        0       1   663       1   663   0.000000e+00      1269  \n",
      "492        0       1   157     226   382   4.836000e-97       314  \n",
      "\n",
      "[493 rows x 12 columns]\n"
     ]
    }
   ],
   "source": [
    "best_matches = pd.DataFrame()\n",
    "\n",
    "bat_prefixes = ['rf', 'dr', 'ra', 'pv', 'pg', 'pa', 'pk', 'ph', 'pd', 'mm', 'ml', 'mb', 'mo', 'md', 'sb', 'ef', 'aj', 'pn', 'en']\n",
    "\n",
    "for prefix in bat_prefixes:\n",
    "\n",
    "    bat_matches = df[df['sseqid'].str.startswith(prefix)]\n",
    "\n",
    "    best_bat_matches = bat_matches.loc[bat_matches.groupby('qseqid')['pident'].idxmax()]\n",
    "\n",
    "    best_matches = pd.concat([best_matches, best_bat_matches], ignore_index=True)\n",
    "\n",
    "\n",
    "print(best_matches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_matches.to_csv('batstring_bestmatches.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "qseqid\n",
      "59463.ENSMLUP00000000083    17\n",
      "59463.ENSMLUP00000001427    17\n",
      "59463.ENSMLUP00000001789    17\n",
      "59463.ENSMLUP00000002616    17\n",
      "59463.ENSMLUP00000002847    17\n",
      "59463.ENSMLUP00000003315    17\n",
      "59463.ENSMLUP00000003322    17\n",
      "59463.ENSMLUP00000004009    17\n",
      "59463.ENSMLUP00000004202    17\n",
      "59463.ENSMLUP00000005054    17\n",
      "59463.ENSMLUP00000005277    17\n",
      "59463.ENSMLUP00000005366    17\n",
      "59463.ENSMLUP00000005595    17\n",
      "59463.ENSMLUP00000008027    17\n",
      "59463.ENSMLUP00000009180    17\n",
      "59463.ENSMLUP00000009404    17\n",
      "59463.ENSMLUP00000010380    17\n",
      "59463.ENSMLUP00000011160    17\n",
      "59463.ENSMLUP00000011196    17\n",
      "59463.ENSMLUP00000011227    17\n",
      "59463.ENSMLUP00000011717    17\n",
      "59463.ENSMLUP00000012090    17\n",
      "59463.ENSMLUP00000013453    17\n",
      "59463.ENSMLUP00000013787    17\n",
      "59463.ENSMLUP00000014213    17\n",
      "59463.ENSMLUP00000016071    17\n",
      "59463.ENSMLUP00000016363    17\n",
      "59463.ENSMLUP00000016529    17\n",
      "59463.ENSMLUP00000022559    17\n",
      "Name: bat_prefix, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "best_matches['bat_prefix'] = best_matches['sseqid'].str[:2]\n",
    "\n",
    "# Group by 'qseqid' and count unique 'bat_prefix' values for each human protein\n",
    "bat_counts_per_human = best_matches.groupby('qseqid')['bat_prefix'].nunique()\n",
    "\n",
    "# Display the counts\n",
    "print(bat_counts_per_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "qseqids with less than 17 bat matches: []\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "qseqids_less_than = bat_counts_per_human[bat_counts_per_human < 17].index.tolist()\n",
    "print(\"qseqids with less than 17 bat matches:\", qseqids_less_than)\n",
    "print(len(qseqids_less_than))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.933"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_matches['pident'].median()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PAIRED FASTA GENERATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bat-only paired sequences for 59463.ENSMLUP00000000083 written to batstring_paired\\59463.ENSMLUP00000000083_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000001427 written to batstring_paired\\59463.ENSMLUP00000001427_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000001789 written to batstring_paired\\59463.ENSMLUP00000001789_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000002616 written to batstring_paired\\59463.ENSMLUP00000002616_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000002847 written to batstring_paired\\59463.ENSMLUP00000002847_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000003315 written to batstring_paired\\59463.ENSMLUP00000003315_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000003322 written to batstring_paired\\59463.ENSMLUP00000003322_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000004009 written to batstring_paired\\59463.ENSMLUP00000004009_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000004202 written to batstring_paired\\59463.ENSMLUP00000004202_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000005054 written to batstring_paired\\59463.ENSMLUP00000005054_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000005277 written to batstring_paired\\59463.ENSMLUP00000005277_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000005366 written to batstring_paired\\59463.ENSMLUP00000005366_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000005595 written to batstring_paired\\59463.ENSMLUP00000005595_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000008027 written to batstring_paired\\59463.ENSMLUP00000008027_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000009180 written to batstring_paired\\59463.ENSMLUP00000009180_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000009404 written to batstring_paired\\59463.ENSMLUP00000009404_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000010380 written to batstring_paired\\59463.ENSMLUP00000010380_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000011160 written to batstring_paired\\59463.ENSMLUP00000011160_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000011196 written to batstring_paired\\59463.ENSMLUP00000011196_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000011227 written to batstring_paired\\59463.ENSMLUP00000011227_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000011717 written to batstring_paired\\59463.ENSMLUP00000011717_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000012090 written to batstring_paired\\59463.ENSMLUP00000012090_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000013453 written to batstring_paired\\59463.ENSMLUP00000013453_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000013787 written to batstring_paired\\59463.ENSMLUP00000013787_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000014213 written to batstring_paired\\59463.ENSMLUP00000014213_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000016071 written to batstring_paired\\59463.ENSMLUP00000016071_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000016363 written to batstring_paired\\59463.ENSMLUP00000016363_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000016529 written to batstring_paired\\59463.ENSMLUP00000016529_bat_only.fasta\n",
      "Bat-only paired sequences for 59463.ENSMLUP00000022559 written to batstring_paired\\59463.ENSMLUP00000022559_bat_only.fasta\n",
      "All bat-only paired sequences written to directory: batstring_paired\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "\n",
    "# File path\n",
    "bat_fasta_file = \"realmergedprots_noen.faa\"  # FASTA file containing bat protein sequences\n",
    "blast_results_file = \"batstring_bestmatches.txt\"  # Should contain the best_matches DataFrame\n",
    "output_dir = \"batstring_paired\"  # Directory to store per-gene FASTA files\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Function to load sequences while handling duplicates\n",
    "def load_fasta_with_duplicates(fasta_file):\n",
    "    sequences = {}\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        if record.id not in sequences:\n",
    "            sequences[record.id] = record\n",
    "        else:\n",
    "            print(f\"Duplicate ID found: {record.id}, ignoring this entry.\")\n",
    "    return sequences\n",
    "\n",
    "# Load bat sequences\n",
    "bat_sequences = load_fasta_with_duplicates(bat_fasta_file)\n",
    "\n",
    "# Load best_matches DataFrame\n",
    "blast_df = pd.read_csv(blast_results_file, sep=\"\\t\")\n",
    "\n",
    "# Process each human gene and its associated bat genes\n",
    "for qseqid, group in blast_df.groupby(\"qseqid\"):\n",
    "    output_file = os.path.join(output_dir, f\"{qseqid}_bat_only.fasta\")\n",
    "    \n",
    "    with open(output_file, \"w\") as output_fasta:\n",
    "        # Write only aligned bat protein sequences\n",
    "        for _, row in group.iterrows():\n",
    "            sseqid_full = row[\"sseqid\"]\n",
    "            prefix = sseqid_full[:2]\n",
    "            sseqid = sseqid_full\n",
    "\n",
    "            if sseqid in bat_sequences:\n",
    "                sseq = bat_sequences[sseqid].seq\n",
    "                sdesc = bat_sequences[sseqid].description\n",
    "\n",
    "                # Note the ID source based on prefix\n",
    "                if prefix == \"pn\" or prefix == 'en':\n",
    "                    id_source = \"GenBank\"\n",
    "                else:\n",
    "                    id_source = \"RefSeq\"\n",
    "\n",
    "                # Write the bat protein sequence\n",
    "                output_fasta.write(f\">{sseqid} {sdesc} [{id_source}]\\n{sseq}\\n\")\n",
    "            else:\n",
    "                print(f\"SseqID {sseqid} not found in bat proteome.\")\n",
    "\n",
    "    print(f\"Bat-only paired sequences for {qseqid} written to {output_file}\")\n",
    "\n",
    "print(f\"All bat-only paired sequences written to directory: {output_dir}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bat-only paired sequences for P00558 written to housekeeping_paired\\P00558_bat_only.fasta\n",
      "Bat-only paired sequences for P04406 written to housekeeping_paired\\P04406_bat_only.fasta\n",
      "Bat-only paired sequences for P05388 written to housekeeping_paired\\P05388_bat_only.fasta\n",
      "Bat-only paired sequences for P60709 written to housekeeping_paired\\P60709_bat_only.fasta\n",
      "Bat-only paired sequences for P62277 written to housekeeping_paired\\P62277_bat_only.fasta\n",
      "All bat-only paired sequences written to directory: housekeeping_paired\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# File paths\n",
    "bat_fasta_file = \"realmergedprots_filtered.faa\"  # FASTA file containing bat protein sequences\n",
    "blast_results_file = \"housekeeping_bestmatches.txt\"  # Should contain the best_matches DataFrame\n",
    "output_dir = \"housekeeping_paired\"  # Directory to store per-gene FASTA files\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Function to load sequences while handling duplicates\n",
    "def load_fasta_with_duplicates(fasta_file):\n",
    "    sequences = {}\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        if record.id not in sequences:\n",
    "            sequences[record.id] = record\n",
    "        else:\n",
    "            print(f\"Duplicate ID found: {record.id}, ignoring this entry.\")\n",
    "    return sequences\n",
    "\n",
    "# Load bat sequences\n",
    "bat_sequences = load_fasta_with_duplicates(bat_fasta_file)\n",
    "\n",
    "# Load best_matches DataFrame\n",
    "blast_df = pd.read_csv(blast_results_file, sep=\"\\t\")\n",
    "\n",
    "# Process each human gene and its associated bat genes\n",
    "for qseqid, group in blast_df.groupby(\"qseqid\"):\n",
    "    output_file = os.path.join(output_dir, f\"{qseqid}_bat_only.fasta\")\n",
    "    \n",
    "    with open(output_file, \"w\") as output_fasta:\n",
    "        # Write only aligned bat protein sequences\n",
    "        for _, row in group.iterrows():\n",
    "            sseqid = row[\"sseqid\"]  # Use sseqid directly as the key for lookup\n",
    "\n",
    "            if sseqid in bat_sequences:\n",
    "                sseq = bat_sequences[sseqid].seq\n",
    "                sdesc = bat_sequences[sseqid].description\n",
    "\n",
    "                # Write the bat protein sequence without altering the label\n",
    "                output_fasta.write(f\">{sseqid} {sdesc}\\n{sseq}\\n\")\n",
    "            else:\n",
    "                print(f\"SseqID {sseqid} not found in bat proteome.\")\n",
    "\n",
    "    print(f\"Bat-only paired sequences for {qseqid} written to {output_file}\")\n",
    "\n",
    "print(f\"All bat-only paired sequences written to directory: {output_dir}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then MSA in command line (MAFFT) OPTIONAL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nucleotide file generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading nucleotide sequences from mergedprotscds.fasta...\n",
      "Loaded 862310 nucleotide sequences.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000000083_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000000083_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000000083_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000001427_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000001427_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000001427_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000001789_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000001789_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000001789_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000002616_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000002616_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000002616_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000002847_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000002847_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000002847_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000003315_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000003315_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000003315_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000003322_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000003322_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000003322_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000004009_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000004009_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000004009_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000004202_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000004202_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000004202_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000005054_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000005054_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000005054_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000005277_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000005277_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000005277_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000005366_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000005366_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000005366_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000005595_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000005595_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000005595_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000008027_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000008027_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000008027_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000009180_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000009180_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000009180_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000009404_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000009404_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000009404_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000010380_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000010380_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000010380_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000011160_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000011160_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000011160_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000011196_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000011196_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000011196_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000011227_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000011227_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000011227_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000011717_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000011717_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000011717_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000012090_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000012090_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000012090_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000013453_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000013453_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000013453_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000013787_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000013787_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000013787_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000014213_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000014213_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000014213_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000016071_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000016071_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000016071_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000016363_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000016363_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000016363_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000016529_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000016529_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000016529_bat_only_codon_alignment.fasta.\n",
      "Extracted nucleotide sequences for 59463.ENSMLUP00000022559_bat_only.\n",
      "Running pal2nal for 59463.ENSMLUP00000022559_bat_only...\n",
      "Codon alignment saved to batstring_nucs\\59463.ENSMLUP00000022559_bat_only_codon_alignment.fasta.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "# File paths\n",
    "nucleotide_fasta = \"mergedprotscds.fasta\"  # Large file with all nucleotide sequences\n",
    "msa_input_dir = \"batstring_paired\"  # Directory with per-gene protein alignments\n",
    "output_dir = \"batstring_nucs\"  # Directory to store codon alignments\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Function to parse IDs and handle repeated names\n",
    "def extract_gene_id(header):\n",
    "    \"\"\"Extract the first gene ID from a header with duplicate IDs.\"\"\"\n",
    "    return header.split()[0]  # Take the first part of the header\n",
    "\n",
    "# Load nucleotide sequences into a dictionary, using only the first part of the ID\n",
    "print(\"Loading nucleotide sequences from mergedprotscds.fasta...\")\n",
    "nucleotide_dict = {}\n",
    "for record in SeqIO.parse(nucleotide_fasta, \"fasta\"):\n",
    "    gene_id = extract_gene_id(record.id)\n",
    "    nucleotide_dict[gene_id] = record\n",
    "print(f\"Loaded {len(nucleotide_dict)} nucleotide sequences.\")\n",
    "\n",
    "# Process each MSA input file\n",
    "for msa_file in os.listdir(msa_input_dir):\n",
    "    if msa_file.endswith(\".fasta\") or msa_file.endswith(\".fa\"):\n",
    "        msa_path = os.path.join(msa_input_dir, msa_file)\n",
    "        gene_name = os.path.splitext(msa_file)[0]\n",
    "        output_nucleotide_file = os.path.join(output_dir, f\"{gene_name}_nucleotide.fasta\")\n",
    "        output_codon_alignment_file = os.path.join(output_dir, f\"{gene_name}_codon_alignment.fasta\")\n",
    "\n",
    "        # Extract protein IDs from the MSA file\n",
    "        protein_ids = []\n",
    "        with open(msa_path, \"r\") as msa_handle:\n",
    "            for record in SeqIO.parse(msa_handle, \"fasta\"):\n",
    "                protein_id = extract_gene_id(record.id)  # Extract the first part of the header\n",
    "                protein_ids.append(protein_id)\n",
    "        \n",
    "        # Match nucleotide sequences to the protein IDs\n",
    "        matching_nucleotides = []\n",
    "        for pid in protein_ids:\n",
    "            if pid in nucleotide_dict:\n",
    "                matching_nucleotides.append(nucleotide_dict[pid])\n",
    "            else:\n",
    "                print(f\"Warning: Protein ID {pid} not found in nucleotide file.\")\n",
    "        \n",
    "        # Write matching nucleotide sequences to a new file\n",
    "        if matching_nucleotides:\n",
    "            with open(output_nucleotide_file, \"w\") as output_handle:\n",
    "                SeqIO.write(matching_nucleotides, output_handle, \"fasta\")\n",
    "            print(f\"Extracted nucleotide sequences for {gene_name}.\")\n",
    "\n",
    "            # Run pal2nal\n",
    "            print(f\"Running pal2nal for {gene_name}...\")\n",
    "            pal2nal_command = f\"perl pal2nal.pl {msa_path} {output_nucleotide_file} -output fasta -verbose -nogap > {output_codon_alignment_file}\"\n",
    "            os.system(pal2nal_command)\n",
    "            print(f\"Codon alignment saved to {output_codon_alignment_file}.\")\n",
    "        else:\n",
    "            print(f\"No matching nucleotide sequences found for {gene_name}. Skipping...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then MASCE then PAL2NAL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TREES GENERATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing: 59463.ENSMLUP00000000083_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000000083_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032946134.1\n",
      "  Pipistrellus_nathusii: pnCAK6438137.1\n",
      "  Pipistrellus_kuhlii: pkXP_036306523.1\n",
      "  Molossus_molossus: moXP_036123910.1\n",
      "  Myotis_daubentonii: mdXP_059526062.1\n",
      "  Myotis_myotis: mmXP_036195761.1\n",
      "  Saccopteryx_bilineata: sbXP_066093091.1\n",
      "  Desmodus_rotundus: drXP_053766493.1\n",
      "  Artibeus_jamaicensis: ajXP_036983857.1\n",
      "  Phyllostomus_hastatus: phXP_045691911.1\n",
      "  Phyllostomus_discolor: pdXP_028375016.1\n",
      "  Rousettus_aegyptiacus: raXP_015999162.1\n",
      "  Pteropus_giganteus: pgXP_039715138.1\n",
      "  Pteropus_alecto: paXP_006912343.1\n",
      "  Pteropus_vampyrus: pvXP_011366803.1\n",
      "  Myotis_lucifugus: mlXP_023612723.1\n",
      "  Myotis_brandtii: mbXP_005875843.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000000083_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000001427_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000001427_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Desmodus_rotundus: drXP_024433875.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032943217.1\n",
      "  Pteropus_vampyrus: pvXP_023381416.1\n",
      "  Pteropus_giganteus: pgXP_039706701.1\n",
      "  Pteropus_alecto: paXP_015455545.1\n",
      "  Pipistrellus_nathusii: pnCAK6441771.1\n",
      "  Pipistrellus_kuhlii: pkXP_036302719.1\n",
      "  Myotis_myotis: mmXP_036208258.1\n",
      "  Myotis_lucifugus: mlXP_006084811.1\n",
      "  Myotis_brandtii: mbXP_005873252.1\n",
      "  Myotis_daubentonii: mdXP_059543434.1\n",
      "  Molossus_molossus: moXP_036099218.1\n",
      "  Rousettus_aegyptiacus: raXP_015990781.2\n",
      "  Saccopteryx_bilineata: sbXP_066097555.1\n",
      "  Artibeus_jamaicensis: ajXP_036993059.2\n",
      "  Phyllostomus_hastatus: phXP_045707315.1\n",
      "  Phyllostomus_discolor: pdXP_028360749.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000001427_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000001789_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000001789_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032971332.1\n",
      "  Saccopteryx_bilineata: sbXP_066126286.1\n",
      "  Pipistrellus_kuhlii: pkXP_036302572.1\n",
      "  Pipistrellus_nathusii: pnCAK6440718.1\n",
      "  Myotis_myotis: mmXP_036207606.1\n",
      "  Myotis_lucifugus: mlXP_006094076.1\n",
      "  Myotis_daubentonii: mdXP_059546891.1\n",
      "  Myotis_brandtii: mbXP_005859738.1\n",
      "  Molossus_molossus: moXP_036112964.1\n",
      "  Desmodus_rotundus: drXP_024410096.1\n",
      "  Artibeus_jamaicensis: ajXP_036991776.2\n",
      "  Phyllostomus_hastatus: phXP_045715589.1\n",
      "  Phyllostomus_discolor: pdXP_035881363.1\n",
      "  Rousettus_aegyptiacus: raXP_016004729.2\n",
      "  Pteropus_giganteus: pgXP_039710232.1\n",
      "  Pteropus_alecto: paXP_006924536.1\n",
      "  Pteropus_vampyrus: pvXP_011356324.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000001789_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000002616_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000002616_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032990906.1\n",
      "  Desmodus_rotundus: drXP_024425024.1\n",
      "  Artibeus_jamaicensis: ajXP_037018247.2\n",
      "  Phyllostomus_hastatus: phXP_045687553.1\n",
      "  Phyllostomus_discolor: pdXP_028375516.1\n",
      "  Rousettus_aegyptiacus: raXP_016014391.1\n",
      "  Pteropus_vampyrus: pvXP_023385989.1\n",
      "  Pteropus_giganteus: pgXP_039731941.1\n",
      "  Pteropus_alecto: paXP_024898165.1\n",
      "  Myotis_myotis: mmXP_036169772.1\n",
      "  Myotis_daubentonii: mdXP_059553879.1\n",
      "  Myotis_lucifugus: mlXP_014311180.1\n",
      "  Myotis_brandtii: mbXP_005871463.1\n",
      "  Pipistrellus_nathusii: pnCAK6445899.1\n",
      "  Pipistrellus_kuhlii: pkXP_045432385.1\n",
      "  Molossus_molossus: moXP_036096354.1\n",
      "  Saccopteryx_bilineata: sbXP_066137369.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000002616_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000002847_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000002847_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032969289.1\n",
      "  Rousettus_aegyptiacus: raXP_016009829.1\n",
      "  Pteropus_vampyrus: pvXP_011361195.1\n",
      "  Pteropus_giganteus: pgXP_039713374.1\n",
      "  Pteropus_alecto: paXP_006909323.1\n",
      "  Saccopteryx_bilineata: sbXP_066121018.1\n",
      "  Desmodus_rotundus: drXP_024421388.1\n",
      "  Phyllostomus_hastatus: phXP_045703068.1\n",
      "  Phyllostomus_discolor: pdXP_035882395.1\n",
      "  Artibeus_jamaicensis: ajXP_037008806.2\n",
      "  Myotis_myotis: mmXP_036207802.1\n",
      "  Myotis_daubentonii: mdXP_059544627.1\n",
      "  Myotis_lucifugus: mlXP_014314819.1\n",
      "  Myotis_brandtii: mbXP_005878486.1\n",
      "  Pipistrellus_nathusii: pnCAK6441408.1\n",
      "  Pipistrellus_kuhlii: pkXP_036302573.1\n",
      "  Molossus_molossus: moXP_036112869.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000002847_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000003315_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000003315_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Artibeus_jamaicensis: ajXP_037020870.2\n",
      "  Myotis_lucifugus: mlXP_006102661.1\n",
      "  Myotis_brandtii: mbXP_005878849.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032946413.1\n",
      "  Desmodus_rotundus: drXP_024420605.2\n",
      "  Phyllostomus_hastatus: phXP_045680450.1\n",
      "  Phyllostomus_discolor: pdXP_028390175.1\n",
      "  Rousettus_aegyptiacus: raXP_016010199.2\n",
      "  Pteropus_vampyrus: pvXP_011357230.1\n",
      "  Pteropus_giganteus: pgXP_039703228.1\n",
      "  Pteropus_alecto: paXP_006922427.2\n",
      "  Molossus_molossus: moXP_036125294.1\n",
      "  Pipistrellus_nathusii: pnCAK6437681.1\n",
      "  Pipistrellus_kuhlii: pkXP_036308020.1\n",
      "  Myotis_daubentonii: mdXP_059524654.1\n",
      "  Myotis_myotis: mmXP_036197834.1\n",
      "  Saccopteryx_bilineata: sbXP_066120310.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000003315_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000003322_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000003322_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rousettus_aegyptiacus: raXP_015997976.2\n",
      "  Pteropus_giganteus: pgXP_039727061.1\n",
      "  Pteropus_alecto: paXP_006921000.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032986834.1\n",
      "  Saccopteryx_bilineata: sbXP_066095214.1\n",
      "  Desmodus_rotundus: drXP_045047578.2\n",
      "  Artibeus_jamaicensis: ajXP_053524495.1\n",
      "  Phyllostomus_hastatus: phXP_045710189.1\n",
      "  Phyllostomus_discolor: pdXP_028368820.1\n",
      "  Myotis_myotis: mmXP_036191760.1\n",
      "  Myotis_lucifugus: mlXP_006088216.1\n",
      "  Myotis_daubentonii: mdXP_059517512.1\n",
      "  Myotis_brandtii: mbXP_005874619.1\n",
      "  Pipistrellus_kuhlii: pkXP_036295060.1\n",
      "  Pipistrellus_nathusii: pnCAK6436079.1\n",
      "  Molossus_molossus: moXP_036121405.1\n",
      "  Pteropus_vampyrus: pvXP_011357616.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000003322_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000004009_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000004009_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rousettus_aegyptiacus: raXP_015999463.2\n",
      "  Rhinolophus_ferrumequinum: rfXP_032960936.1\n",
      "  Pteropus_vampyrus: pvXP_011362316.1\n",
      "  Pteropus_alecto: paXP_006916440.1\n",
      "  Pteropus_giganteus: pgXP_039738780.1\n",
      "  Desmodus_rotundus: drXP_024418601.1\n",
      "  Phyllostomus_hastatus: phXP_045677680.1\n",
      "  Phyllostomus_discolor: pdXP_028382052.2\n",
      "  Saccopteryx_bilineata: sbXP_066093548.1\n",
      "  Molossus_molossus: moXP_036106138.1\n",
      "  Myotis_myotis: mmXP_036152966.1\n",
      "  Myotis_daubentonii: mdXP_059537099.1\n",
      "  Myotis_lucifugus: mlXP_023612896.1\n",
      "  Myotis_brandtii: mbXP_014386881.1\n",
      "  Artibeus_jamaicensis: ajXP_036984798.2\n",
      "  Pipistrellus_nathusii: pnCAK6442564.1\n",
      "  Pipistrellus_kuhlii: pkXP_045443435.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000004009_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000004202_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000004202_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032946676.1\n",
      "  Rousettus_aegyptiacus: raXP_016006474.2\n",
      "  Pteropus_giganteus: pgXP_039729753.1\n",
      "  Pteropus_alecto: paXP_006924730.1\n",
      "  Pteropus_vampyrus: pvXP_011381398.1\n",
      "  Desmodus_rotundus: drXP_024428890.2\n",
      "  Phyllostomus_hastatus: phXP_045692366.1\n",
      "  Phyllostomus_discolor: pdXP_028376247.1\n",
      "  Artibeus_jamaicensis: ajXP_036985630.2\n",
      "  Saccopteryx_bilineata: sbXP_066119769.1\n",
      "  Myotis_myotis: mmXP_036197705.1\n",
      "  Myotis_lucifugus: mlXP_006095025.1\n",
      "  Myotis_daubentonii: mdXP_059526981.1\n",
      "  Myotis_brandtii: mbXP_005861908.1\n",
      "  Pipistrellus_nathusii: pnCAK6438415.1\n",
      "  Pipistrellus_kuhlii: pkXP_036263632.1\n",
      "  Molossus_molossus: moXP_036124041.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000004202_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000005054_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000005054_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032956699.1\n",
      "  Rousettus_aegyptiacus: raXP_016015693.1\n",
      "  Pteropus_giganteus: pgXP_039725816.1\n",
      "  Pteropus_alecto: paXP_006926112.1\n",
      "  Desmodus_rotundus: drXP_053769596.1\n",
      "  Artibeus_jamaicensis: ajXP_037014889.2\n",
      "  Phyllostomus_hastatus: phXP_045699260.1\n",
      "  Phyllostomus_discolor: pdXP_028364317.1\n",
      "  Molossus_molossus: moXP_036104500.1\n",
      "  Pipistrellus_nathusii: pnCAK6446717.1\n",
      "  Pipistrellus_kuhlii: pkXP_036271341.1\n",
      "  Myotis_myotis: mmXP_036172070.1\n",
      "  Myotis_lucifugus: mlXP_006098082.1\n",
      "  Myotis_daubentonii: mdXP_059556772.1\n",
      "  Saccopteryx_bilineata: sbXP_066103763.1\n",
      "  Myotis_brandtii: mbXP_005883885.2\n",
      "  Pteropus_vampyrus: pvXP_023386725.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000005054_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Skipping empty file: 59463.ENSMLUP00000005277_bat_only_nucleotide_codon_aligned.fasta\n",
      "\n",
      "Processing: 59463.ENSMLUP00000005366_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000005366_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Phyllostomus_discolor: pdXP_035867530.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032960851.1\n",
      "  Pteropus_vampyrus: pvXP_011371577.1\n",
      "  Molossus_molossus: moXP_036105838.1\n",
      "  Artibeus_jamaicensis: ajXP_053515842.1\n",
      "  Rousettus_aegyptiacus: raXP_036084611.1\n",
      "  Pteropus_giganteus: pgXP_039702365.1\n",
      "  Pteropus_alecto: paXP_006916392.1\n",
      "  Pipistrellus_nathusii: pnCAK6442623.1\n",
      "  Pipistrellus_kuhlii: pkXP_036314135.1\n",
      "  Myotis_daubentonii: mdXP_059541456.1\n",
      "  Myotis_myotis: mmXP_036154392.1\n",
      "  Myotis_lucifugus: mlXP_023607616.1\n",
      "  Myotis_brandtii: mbXP_014393357.1\n",
      "  Saccopteryx_bilineata: sbXP_066091206.1\n",
      "  Desmodus_rotundus: drXP_053772323.1\n",
      "  Phyllostomus_hastatus: phXP_045677464.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000005366_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000005595_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000005595_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032946676.1\n",
      "  Rousettus_aegyptiacus: raXP_016006474.2\n",
      "  Pteropus_giganteus: pgXP_039729753.1\n",
      "  Pteropus_alecto: paXP_006924730.1\n",
      "  Pteropus_vampyrus: pvXP_011381398.1\n",
      "  Desmodus_rotundus: drXP_024428890.2\n",
      "  Phyllostomus_hastatus: phXP_045692366.1\n",
      "  Phyllostomus_discolor: pdXP_028376247.1\n",
      "  Artibeus_jamaicensis: ajXP_036985630.2\n",
      "  Saccopteryx_bilineata: sbXP_066119769.1\n",
      "  Myotis_myotis: mmXP_036197705.1\n",
      "  Myotis_lucifugus: mlXP_006095025.1\n",
      "  Myotis_daubentonii: mdXP_059526981.1\n",
      "  Myotis_brandtii: mbXP_005861908.1\n",
      "  Pipistrellus_nathusii: pnCAK6438415.1\n",
      "  Pipistrellus_kuhlii: pkXP_036263632.1\n",
      "  Molossus_molossus: moXP_036124041.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000005595_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000008027_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000008027_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032943586.1\n",
      "  Desmodus_rotundus: drXP_024436450.2\n",
      "  Rousettus_aegyptiacus: raXP_016003904.1\n",
      "  Phyllostomus_hastatus: phXP_045705558.1\n",
      "  Phyllostomus_discolor: pdXP_028380653.1\n",
      "  Artibeus_jamaicensis: ajXP_037021015.2\n",
      "  Molossus_molossus: moXP_036101281.1\n",
      "  Saccopteryx_bilineata: sbXP_066103492.1\n",
      "  Pteropus_vampyrus: pvXP_011359430.1\n",
      "  Pteropus_alecto: paXP_006919993.1\n",
      "  Pteropus_giganteus: pgXP_039698376.1\n",
      "  Myotis_brandtii: mbXP_005865805.2\n",
      "  Myotis_myotis: mmXP_036179611.1\n",
      "  Myotis_daubentonii: mdXP_059563204.1\n",
      "  Pipistrellus_nathusii: pnCAK6447973.1\n",
      "  Pipistrellus_kuhlii: pkXP_036273992.1\n",
      "  Myotis_lucifugus: mlXP_006109748.2\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000008027_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000009180_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000009180_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Myotis_lucifugus: mlXP_023600738.1\n",
      "  Pteropus_vampyrus: pvXP_023384443.1\n",
      "  Saccopteryx_bilineata: sbXP_066107369.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032977584.1\n",
      "  Rousettus_aegyptiacus: raXP_015978729.1\n",
      "  Pteropus_alecto: paXP_006908094.1\n",
      "  Pteropus_giganteus: pgXP_039722978.1\n",
      "  Molossus_molossus: moXP_036113887.1\n",
      "  Desmodus_rotundus: drXP_024414685.1\n",
      "  Artibeus_jamaicensis: ajXP_053529084.1\n",
      "  Phyllostomus_hastatus: phXP_045681526.1\n",
      "  Phyllostomus_discolor: pdXP_028371519.1\n",
      "  Myotis_myotis: mmXP_036182083.1\n",
      "  Myotis_brandtii: mbXP_005864486.1\n",
      "  Myotis_daubentonii: mdXP_059565645.1\n",
      "  Pipistrellus_kuhlii: pkXP_036284360.1\n",
      "  Pipistrellus_nathusii: pnCAK6449506.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000009180_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000009404_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000009404_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032948995.1\n",
      "  Myotis_myotis: mmXP_036199724.1\n",
      "  Phyllostomus_hastatus: phXP_045697522.1\n",
      "  Artibeus_jamaicensis: ajXP_053523998.1\n",
      "  Phyllostomus_discolor: pdXP_028386799.1\n",
      "  Saccopteryx_bilineata: sbXP_066111249.1\n",
      "  Molossus_molossus: moXP_036129501.1\n",
      "  Myotis_daubentonii: mdXP_059531184.1\n",
      "  Desmodus_rotundus: drXP_024427405.1\n",
      "  Rousettus_aegyptiacus: raXP_015996560.2\n",
      "  Pteropus_vampyrus: pvXP_011371263.1\n",
      "  Pteropus_giganteus: pgXP_039737823.1\n",
      "  Pteropus_alecto: paXP_006922897.1\n",
      "  Pipistrellus_nathusii: pnCAK6439716.1\n",
      "  Pipistrellus_kuhlii: pkXP_036298830.1\n",
      "  Myotis_lucifugus: mlXP_006094908.1\n",
      "  Myotis_brandtii: mbXP_005885346.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000009404_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Skipping empty file: 59463.ENSMLUP00000010380_bat_only_nucleotide_codon_aligned.fasta\n",
      "\n",
      "Processing: 59463.ENSMLUP00000011160_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000011160_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032990303.1\n",
      "  Rousettus_aegyptiacus: raXP_015987888.2\n",
      "  Pteropus_vampyrus: pvXP_011365053.1\n",
      "  Pteropus_giganteus: pgXP_039723858.1\n",
      "  Pteropus_alecto: paXP_006904386.1\n",
      "  Myotis_brandtii: mbXP_005873961.2\n",
      "  Myotis_myotis: mmXP_036167606.1\n",
      "  Myotis_daubentonii: mdXP_059552122.1\n",
      "  Myotis_lucifugus: mlXP_006100443.1\n",
      "  Pipistrellus_kuhlii: pkXP_036305808.1\n",
      "  Pipistrellus_nathusii: pnCAK6445298.1\n",
      "  Molossus_molossus: moXP_036095710.1\n",
      "  Saccopteryx_bilineata: sbXP_066125039.1\n",
      "  Desmodus_rotundus: drXP_024416550.1\n",
      "  Artibeus_jamaicensis: ajXP_036998032.2\n",
      "  Phyllostomus_hastatus: phXP_045709182.1\n",
      "  Phyllostomus_discolor: pdXP_028376370.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000011160_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000011196_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000011196_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Pipistrellus_kuhlii: pkXP_036272319.2\n",
      "  Rhinolophus_ferrumequinum: rfXP_032984803.1\n",
      "  Pipistrellus_nathusii: pnCAK6437196.1\n",
      "  Myotis_daubentonii: mdXP_059522145.1\n",
      "  Myotis_myotis: mmXP_036201853.1\n",
      "  Desmodus_rotundus: drXP_024425231.1\n",
      "  Phyllostomus_hastatus: phXP_045701093.1\n",
      "  Artibeus_jamaicensis: ajXP_037003923.1\n",
      "  Phyllostomus_discolor: pdXP_028385790.1\n",
      "  Molossus_molossus: moXP_036130990.1\n",
      "  Pteropus_giganteus: pgXP_039722132.1\n",
      "  Pteropus_alecto: paXP_024902552.1\n",
      "  Saccopteryx_bilineata: sbXP_066127896.1\n",
      "  Myotis_brandtii: mbXP_005863442.1\n",
      "  Rousettus_aegyptiacus: raXP_036088371.1\n",
      "  Pteropus_vampyrus: pvXP_023391126.1\n",
      "  Myotis_lucifugus: mlXP_023604919.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000011196_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000011227_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000011227_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Myotis_brandtii: mbXP_014392858.1\n",
      "  Myotis_myotis: mmXP_036201301.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032983309.1\n",
      "  Saccopteryx_bilineata: sbXP_066127918.1\n",
      "  Rousettus_aegyptiacus: raXP_036088415.1\n",
      "  Pteropus_vampyrus: pvXP_011366253.1\n",
      "  Pteropus_alecto: paXP_015444810.1\n",
      "  Pteropus_giganteus: pgXP_039722139.1\n",
      "  Molossus_molossus: moXP_036131859.1\n",
      "  Pipistrellus_nathusii: pnCAK6437199.1\n",
      "  Pipistrellus_kuhlii: pkXP_036272368.1\n",
      "  Myotis_daubentonii: mdXP_059522134.1\n",
      "  Myotis_lucifugus: mlXP_014304701.2\n",
      "  Desmodus_rotundus: drXP_024425273.2\n",
      "  Phyllostomus_hastatus: phXP_045701187.1\n",
      "  Phyllostomus_discolor: pdXP_028385851.1\n",
      "  Artibeus_jamaicensis: ajXP_037003912.2\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000011227_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000011717_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000011717_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Pipistrellus_nathusii: pnCAK6438461.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032945859.1\n",
      "  Pteropus_vampyrus: pvXP_011379562.1\n",
      "  Phyllostomus_hastatus: phXP_045693148.1\n",
      "  Desmodus_rotundus: drXP_024409255.1\n",
      "  Molossus_molossus: moXP_036124284.1\n",
      "  Rousettus_aegyptiacus: raXP_016006487.2\n",
      "  Pteropus_giganteus: pgXP_039729736.1\n",
      "  Pteropus_alecto: paXP_006924699.1\n",
      "  Phyllostomus_discolor: pdXP_028375175.2\n",
      "  Saccopteryx_bilineata: sbXP_066119747.1\n",
      "  Pipistrellus_kuhlii: pkXP_036263611.1\n",
      "  Myotis_myotis: mmXP_036194923.1\n",
      "  Myotis_daubentonii: mdXP_059526989.1\n",
      "  Myotis_lucifugus: mlXP_006095031.1\n",
      "  Myotis_brandtii: mbXP_005862025.1\n",
      "  Artibeus_jamaicensis: ajXP_036985597.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000011717_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000012090_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000012090_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Rhinolophus_ferrumequinum: rfXP_032955476.1\n",
      "  Artibeus_jamaicensis: ajXP_037016765.2\n",
      "  Myotis_brandtii: mbXP_014387840.1\n",
      "  Saccopteryx_bilineata: sbXP_066094798.1\n",
      "  Molossus_molossus: moXP_036133433.1\n",
      "  Desmodus_rotundus: drXP_045040527.2\n",
      "  Rousettus_aegyptiacus: raXP_016006050.1\n",
      "  Pteropus_vampyrus: pvXP_011365554.1\n",
      "  Pteropus_giganteus: pgXP_039739437.1\n",
      "  Myotis_myotis: mmXP_036211051.1\n",
      "  Myotis_daubentonii: mdXP_059533887.1\n",
      "  Phyllostomus_hastatus: phXP_045673108.1\n",
      "  Phyllostomus_discolor: pdXP_035871977.1\n",
      "  Pipistrellus_nathusii: pnCAK6442310.1\n",
      "  Myotis_lucifugus: mlXP_023604648.1\n",
      "  Pipistrellus_kuhlii: pkXP_036296085.1\n",
      "  Pteropus_alecto: paXP_024900574.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000012090_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000013453_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000013453_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Phyllostomus_hastatus: phXP_045680629.1\n",
      "  Myotis_daubentonii: mdXP_059524875.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032947164.1\n",
      "  Myotis_myotis: mmXP_036197234.1\n",
      "  Myotis_lucifugus: mlXP_006105147.1\n",
      "  Myotis_brandtii: mbXP_014405388.1\n",
      "  Pipistrellus_nathusii: pnCAK6437759.1\n",
      "  Pipistrellus_kuhlii: pkXP_036307678.1\n",
      "  Saccopteryx_bilineata: sbXP_066119007.1\n",
      "  Artibeus_jamaicensis: ajXP_036985111.2\n",
      "  Phyllostomus_discolor: pdXP_035889055.1\n",
      "  Molossus_molossus: moXP_036124931.1\n",
      "  Rousettus_aegyptiacus: raXP_015997557.2\n",
      "  Pteropus_vampyrus: pvXP_023386162.1\n",
      "  Pteropus_giganteus: pgXP_039730598.1\n",
      "  Pteropus_alecto: paXP_006922338.1\n",
      "  Desmodus_rotundus: drXP_045055206.2\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000013453_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000013787_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000013787_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Saccopteryx_bilineata: sbXP_066129585.1\n",
      "  Pipistrellus_kuhlii: pkXP_036280301.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032952375.1\n",
      "  Rousettus_aegyptiacus: raXP_036080377.1\n",
      "  Pteropus_vampyrus: pvXP_023388557.1\n",
      "  Pteropus_giganteus: pgXP_039734888.1\n",
      "  Molossus_molossus: moXP_036097890.1\n",
      "  Desmodus_rotundus: drXP_024432590.2\n",
      "  Artibeus_jamaicensis: ajXP_036993816.1\n",
      "  Phyllostomus_hastatus: phXP_045680737.1\n",
      "  Phyllostomus_discolor: pdXP_028383086.1\n",
      "  Myotis_myotis: mmXP_036168455.1\n",
      "  Myotis_daubentonii: mdXP_059555375.1\n",
      "  Pipistrellus_nathusii: pnCAK6446407.1\n",
      "  Myotis_lucifugus: mlXP_014305940.1\n",
      "  Myotis_brandtii: mbXP_014386464.1\n",
      "  Pteropus_alecto: paXP_024905335.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000013787_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000014213_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000014213_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Pteropus_vampyrus: pvXP_011379860.1\n",
      "  Phyllostomus_hastatus: phXP_045707297.1\n",
      "  Pipistrellus_kuhlii: pkXP_036300663.1\n",
      "  Myotis_daubentonii: mdXP_059543353.1\n",
      "  Myotis_lucifugus: mlXP_006084840.2\n",
      "  Myotis_brandtii: mbXP_005880821.2\n",
      "  Rhinolophus_ferrumequinum: rfXP_032945966.1\n",
      "  Myotis_myotis: mmXP_036203510.1\n",
      "  Pteropus_alecto: paXP_024898058.1\n",
      "  Desmodus_rotundus: drXP_024420519.1\n",
      "  Phyllostomus_discolor: pdXP_028360552.2\n",
      "  Rousettus_aegyptiacus: raXP_015991808.2\n",
      "  Pteropus_giganteus: pgXP_039706604.1\n",
      "  Molossus_molossus: moXP_036098758.1\n",
      "  Saccopteryx_bilineata: sbXP_066097288.1\n",
      "  Artibeus_jamaicensis: ajXP_036987670.1\n",
      "  Pipistrellus_nathusii: pnCAK6441801.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000014213_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "\n",
      "Processing: 59463.ENSMLUP00000016071_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000016071_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Phyllostomus_discolor: pdXP_028369249.1\n",
      "  Rousettus_aegyptiacus: raXP_016004931.2\n",
      "  Pteropus_vampyrus: pvXP_023388381.1\n",
      "  Pteropus_giganteus: pgXP_039698121.1\n",
      "  Pteropus_alecto: paXP_015446900.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032971383.1\n",
      "  Desmodus_rotundus: drXP_024410613.2\n",
      "  Artibeus_jamaicensis: ajXP_037021567.2\n",
      "  Phyllostomus_hastatus: phXP_045715842.1\n",
      "  Saccopteryx_bilineata: sbXP_066126507.1\n",
      "  Molossus_molossus: moXP_036111589.1\n",
      "  Pipistrellus_nathusii: pnCAK6440610.1\n",
      "  Pipistrellus_kuhlii: pkXP_036316803.1\n",
      "  Myotis_myotis: mmXP_036205755.1\n",
      "  Myotis_daubentonii: mdXP_059547208.1\n",
      "  Myotis_lucifugus: mlXP_006104586.1\n",
      "  Myotis_brandtii: mbXP_005873547.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000016071_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Skipping empty file: 59463.ENSMLUP00000016363_bat_only_nucleotide_codon_aligned.fasta\n",
      "\n",
      "Processing: 59463.ENSMLUP00000016529_bat_only_nucleotide_codon_aligned.fasta\n",
      "Extracted sequence names for 59463.ENSMLUP00000016529_bat_only_nucleotide_codon_aligned.fasta:\n",
      "  Saccopteryx_bilineata: sbXP_066111434.1\n",
      "  Desmodus_rotundus: drXP_045055206.2\n",
      "  Myotis_myotis: mmXP_036197236.1\n",
      "  Rhinolophus_ferrumequinum: rfXP_032947163.1\n",
      "  Rousettus_aegyptiacus: raXP_015997557.2\n",
      "  Pteropus_vampyrus: pvXP_023386162.1\n",
      "  Pteropus_giganteus: pgXP_039730598.1\n",
      "  Pteropus_alecto: paXP_006922338.1\n",
      "  Pipistrellus_nathusii: pnCAK6437761.1\n",
      "  Pipistrellus_kuhlii: pkXP_036307681.1\n",
      "  Myotis_brandtii: mbXP_014397415.1\n",
      "  Myotis_daubentonii: mdXP_059524892.1\n",
      "  Myotis_lucifugus: mlXP_014302774.1\n",
      "  Molossus_molossus: moXP_036125124.1\n",
      "  Artibeus_jamaicensis: ajXP_036985111.2\n",
      "  Phyllostomus_discolor: pdXP_035889055.1\n",
      "  Phyllostomus_hastatus: phXP_045680628.1\n",
      "Unique tree saved to: batstring_trees\\59463.ENSMLUP00000016529_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Skipping empty file: 59463.ENSMLUP00000022559_bat_only_nucleotide_codon_aligned.fasta\n"
     ]
    }
   ],
   "source": [
    "from ete3 import Tree\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "# Paths\n",
    "full_tree_file = \"bat_tags_noen.nwk\"  # Path to the full tree file\n",
    "pal2nal_dir = \"batstring_pal2nal\"  # Directory with cleaned codon-aligned files\n",
    "output_tree_dir = \"batstring_trees\"  # Directory to store pruned trees\n",
    "\n",
    "# Species tag to full name mapping\n",
    "tag_to_species = {\n",
    "    \"rf\": \"Rhinolophus_ferrumequinum\",\n",
    "    \"dr\": \"Desmodus_rotundus\",\n",
    "    \"ra\": \"Rousettus_aegyptiacus\",\n",
    "    \"pv\": \"Pteropus_vampyrus\",\n",
    "    \"pg\": \"Pteropus_giganteus\",\n",
    "    \"pa\": \"Pteropus_alecto\",\n",
    "    \"pk\": \"Pipistrellus_kuhlii\",\n",
    "    \"ph\": \"Phyllostomus_hastatus\",\n",
    "    \"pd\": \"Phyllostomus_discolor\",\n",
    "    \"mm\": \"Myotis_myotis\",\n",
    "    \"ml\": \"Myotis_lucifugus\",\n",
    "    \"mb\": \"Myotis_brandtii\",\n",
    "    \"mo\": \"Molossus_molossus\",\n",
    "    \"md\": \"Myotis_daubentonii\",\n",
    "    \"sb\": \"Saccopteryx_bilineata\",\n",
    "    \"aj\": \"Artibeus_jamaicensis\",\n",
    "    \"pn\": \"Pipistrellus_nathusii\",\n",
    "    \"en\": \"Eptesicus_nilssonii\",\n",
    "}\n",
    "\n",
    "# Create output directory if it doesn't exist\n",
    "os.makedirs(output_tree_dir, exist_ok=True)\n",
    "\n",
    "# Load the full tree\n",
    "full_tree = Tree(full_tree_file, format=1)\n",
    "\n",
    "# Process each alignment file\n",
    "for alignment_file in os.listdir(pal2nal_dir):\n",
    "    if alignment_file.endswith(\".fasta\"):\n",
    "        alignment_path = os.path.join(pal2nal_dir, alignment_file)\n",
    "\n",
    "        # Check if the file is empty\n",
    "        if os.stat(alignment_path).st_size == 0:\n",
    "            print(f\"Skipping empty file: {alignment_file}\")\n",
    "            continue\n",
    "\n",
    "        print(f\"\\nProcessing: {alignment_file}\")\n",
    "\n",
    "        # Extract sequence names from the alignment\n",
    "        sequence_names = {}\n",
    "        for record in SeqIO.parse(alignment_path, \"fasta\"):\n",
    "            tag = record.id[:2]  # Extract the first two letters of the record ID\n",
    "            if tag in tag_to_species:\n",
    "                species_name = tag_to_species[tag]\n",
    "                sequence_names[species_name] = record.id\n",
    "            else:\n",
    "                print(f\"Warning: Unrecognized tag '{tag}' in {record.id}\")\n",
    "\n",
    "        # Debugging: Print extracted sequence names\n",
    "        print(f\"Extracted sequence names for {alignment_file}:\")\n",
    "        for species, seq_id in sequence_names.items():\n",
    "            print(f\"  {species}: {seq_id}\")\n",
    "\n",
    "        # Copy the full tree for modification\n",
    "        unique_tree = full_tree.copy()\n",
    "\n",
    "        # Rename leaves to match sequence names and remove unmatched leaves\n",
    "        unmatched_leaves = []\n",
    "        for leaf in unique_tree.get_leaves():\n",
    "            if leaf.name in sequence_names:\n",
    "                leaf.name = sequence_names[leaf.name]  # Rename leaf to match sequence ID\n",
    "            else:\n",
    "                unmatched_leaves.append(leaf.name)\n",
    "\n",
    "        # Debugging: Print unmatched leaves\n",
    "        if unmatched_leaves:\n",
    "            print(f\"Unmatched leaves for {alignment_file}: {unmatched_leaves}\")\n",
    "\n",
    "        # Remove unmatched leaves\n",
    "        for leaf_name in unmatched_leaves:\n",
    "            leaf = unique_tree.search_nodes(name=leaf_name)[0]\n",
    "            leaf.detach()\n",
    "\n",
    "        # Save the unique tree\n",
    "        output_tree_path = os.path.join(output_tree_dir, f\"{alignment_file}_pruned.nwk\")\n",
    "        unique_tree.write(outfile=output_tree_path, format=1)\n",
    "        print(f\"Unique tree saved to: {output_tree_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 59463.ENSMLUP00000000083_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000000083_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000001427_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000001427_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000001789_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000001789_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000002616_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000002616_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000002847_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000002847_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000003315_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000003315_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000003322_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000003322_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000004009_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000004009_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000004202_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000004202_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000005054_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000005054_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000005366_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000005366_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000005595_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000005595_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000008027_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000008027_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000009180_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000009180_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000009404_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000009404_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000011160_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000011160_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000011196_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000011196_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000011227_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000011227_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000011717_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000011717_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000012090_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000012090_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000013453_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000013453_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000013787_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000013787_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000014213_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000014213_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000016071_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000016071_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n",
      "Processed 59463.ENSMLUP00000016529_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk -> batstring_trees_labeled\\59463.ENSMLUP00000016529_bat_only_nucleotide_codon_aligned.fasta_pruned.nwk\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import os\n",
    "import re\n",
    "\n",
    "SOURCE_DIR = 'batstring_trees'\n",
    "DEST_DIR = 'batstring_trees_labeled'\n",
    "\n",
    "os.makedirs(DEST_DIR, exist_ok=True)\n",
    "\n",
    "# Prefixes that should be labeled as {Foreground}\n",
    "foreground_prefixes = ['mm', 'ml', 'md', 'mb', 'dr', 'rf']\n",
    "\n",
    "# Regex: capture (1) node/tip name, (2) colon+branch_length\n",
    "pattern = re.compile(r'([^\\(\\),:]+)(:[0-9\\.eE\\-\\+]+)')\n",
    "\n",
    "def replacer(match):\n",
    "    \"\"\"Decide {Foreground} vs. {Background}.\"\"\"\n",
    "    taxon_name = match.group(1)\n",
    "    branch_length = match.group(2)\n",
    "    if any(taxon_name.startswith(pref) for pref in foreground_prefixes):\n",
    "        return f\"{taxon_name}{{Foreground}}{branch_length}\"\n",
    "    else:\n",
    "        return f\"{taxon_name}{{Background}}{branch_length}\"\n",
    "\n",
    "for filename in os.listdir(SOURCE_DIR):\n",
    "    if filename.endswith('.nwk'):\n",
    "        source_file = os.path.join(SOURCE_DIR, filename)\n",
    "        dest_file = os.path.join(DEST_DIR, filename)\n",
    "\n",
    "        with open(source_file, 'r') as infile:\n",
    "            data = infile.read()\n",
    "\n",
    "        # --- FORCE ONE LINE ---\n",
    "        # Remove all linebreaks:\n",
    "        data = data.replace('\\n', '').replace('\\r', '')\n",
    "        # (Optional) remove extra spaces if you like:\n",
    "        # data = ' '.join(data.split())\n",
    "\n",
    "        # Perform the labeling\n",
    "        data_labeled = pattern.sub(replacer, data)\n",
    "\n",
    "        # Write out WITHOUT adding an extra newline at the end\n",
    "        with open(dest_file, 'w') as outfile:\n",
    "            outfile.write(data_labeled)\n",
    "\n",
    "        print(f\"Processed {filename} -> {dest_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DN DS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59463.ENSMLUP00000000083: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000001427: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000001789: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000002616: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000002847: p-value 0.4593 >= 0.05, not added\n",
      "59463.ENSMLUP00000003315: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000003322: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000004009: p-value 0.3741 >= 0.05, not added\n",
      "59463.ENSMLUP00000004202: p-value 0.2994 >= 0.05, not added\n",
      "59463.ENSMLUP00000005054: 'Likelihood ratio test' line not found in 59463.ENSMLUP00000005054_bat_only_nucleotide_codon_aligned.fasta_busted.log\n",
      "59463.ENSMLUP00000005366: p-value 0.2931 >= 0.05, not added\n",
      "59463.ENSMLUP00000005595: p-value 0.2989 >= 0.05, not added\n",
      "59463.ENSMLUP00000008027: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000009180: p-value 0.5 >= 0.05, not added\n",
      "59463.ENSMLUP00000009404: 'Likelihood ratio test' line not found in 59463.ENSMLUP00000009404_bat_only_nucleotide_codon_aligned.fasta_busted.log\n",
      "59463.ENSMLUP00000011160: 'Likelihood ratio test' line not found in 59463.ENSMLUP00000011160_bat_only_nucleotide_codon_aligned.fasta_busted.log\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "# Define the directory containing your .log files\n",
    "log_dir = 'batstring_busted'  # Update this if your directory name is different\n",
    "\n",
    "# Output file to store the protein names\n",
    "output_file = 'batstring_bustede1_prots_nofdr.txt'\n",
    "\n",
    "# Open the output file in write mode\n",
    "with open(output_file, 'w') as out_f:\n",
    "\n",
    "    # Iterate over all .log files in the directory\n",
    "    for filename in os.listdir(log_dir):\n",
    "        if filename.endswith('.log'):\n",
    "            filepath = os.path.join(log_dir, filename)\n",
    "            # Extract the protein name from the filename based on the rules\n",
    "            # Split the filename by '_'\n",
    "            parts = filename.split('_')\n",
    "            if parts[0].startswith('NP') or parts[0].startswith('XP'):\n",
    "                # If the protein name starts with 'NP' or 'XP', join the first two parts\n",
    "                protein_name = '_'.join(parts[:2])\n",
    "            else:\n",
    "                # Otherwise, the protein name is the first part\n",
    "                protein_name = parts[0]\n",
    "\n",
    "            # Open and read the log file\n",
    "            with open(filepath, 'r') as log_f:\n",
    "                lines = log_f.readlines()\n",
    "                for line in lines:\n",
    "                    if 'Likelihood ratio test for episodic diversifying positive selection' in line:\n",
    "                        # Extract the p-value from this line\n",
    "                        match = re.search(r'p =\\s*([\\d\\.eE+-]+)', line)\n",
    "                        if match:\n",
    "                            p_value_str = match.group(1)\n",
    "                            try:\n",
    "                                p_value = float(p_value_str)\n",
    "                                if p_value < 0.05:\n",
    "                                    # Write the protein name to the output file\n",
    "                                    out_f.write(protein_name + '\\n')\n",
    "                                    print(f\"{protein_name}: p-value {p_value} < 0.05, added to {output_file}\")\n",
    "                                else:\n",
    "                                    print(f\"{protein_name}: p-value {p_value} >= 0.05, not added\")\n",
    "                            except ValueError:\n",
    "                                print(f\"{protein_name}: Could not convert p-value '{p_value_str}' to float\")\n",
    "                        else:\n",
    "                            print(f\"{protein_name}: p-value not found in line: {line.strip()}\")\n",
    "                        # Break after processing the line\n",
    "                        break\n",
    "                else:\n",
    "                    # If we didn't find the line with 'Likelihood ratio test...'\n",
    "                    print(f\"{protein_name}: 'Likelihood ratio test' line not found in {filename}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_001159718.1: 'Likelihood ratio test' line not found or no p-value in NP_001159718.1_aligned_codon_alignment.fasta_busted.log\n",
      "NP_036597.1: 'Likelihood ratio test' line not found or no p-value in NP_036597.1_aligned_codon_alignment.fasta_busted.log\n",
      "NP_079279.3: 'Likelihood ratio test' line not found or no p-value in NP_079279.3_aligned_codon_alignment.fasta_busted.log\n",
      "NP_000005.2: raw p = 0.4814, FDR p = 0.5 (not significant)\n",
      "NP_000020.1: raw p = 0.2562, FDR p = 0.5 (not significant)\n",
      "NP_000053.2: raw p = 0.1706, FDR p = 0.5 (not significant)\n",
      "NP_000090.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000091.1: raw p = 0.129, FDR p = 0.5 (not significant)\n",
      "NP_000120.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000122.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000124.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000176.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000286.3: raw p = 0.0754, FDR p = 0.5 (not significant)\n",
      "NP_000292.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000293.2: raw p = 0.4834, FDR p = 0.5 (not significant)\n",
      "NP_000299.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000345.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000350.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000353.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000387.1: raw p = 0.4226, FDR p = 0.5 (not significant)\n",
      "NP_000403.1: raw p = 0.0017, FDR p = 0.07735 (not significant)\n",
      "NP_000479.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000495.1: raw p = 0.0262, FDR p = 0.431 (not significant)\n",
      "NP_000497.1: raw p = 0.3884, FDR p = 0.5 (not significant)\n",
      "NP_000593.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000615.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000884.1: raw p = 0, FDR p = 0 (SIGNIFICANT)\n",
      "NP_000908.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000921.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000925.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_000926.2: raw p = 0.1918, FDR p = 0.5 (not significant)\n",
      "NP_001001851.1: raw p = 0.4268, FDR p = 0.5 (not significant)\n",
      "NP_001001991.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001002235.1: raw p = 0.0748, FDR p = 0.5 (not significant)\n",
      "NP_001002236.1: raw p = 0.0748, FDR p = 0.5 (not significant)\n",
      "NP_001008693.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001012303.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001017962.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001017973.1: raw p = 0.1039, FDR p = 0.5 (not significant)\n",
      "NP_001017974.1: raw p = 0.1003, FDR p = 0.5 (not significant)\n",
      "NP_001027018.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001027019.1: raw p = 0.4578, FDR p = 0.5 (not significant)\n",
      "NP_001027466.1: raw p = 0.1733, FDR p = 0.5 (not significant)\n",
      "NP_001035237.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001035362.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001035983.1: raw p = 0.4406, FDR p = 0.5 (not significant)\n",
      "NP_001075.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001092010.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001094077.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001095886.1: raw p = 0.0296, FDR p = 0.431 (not significant)\n",
      "NP_001100.3: raw p = 0.4065, FDR p = 0.5 (not significant)\n",
      "NP_001101.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001107645.1: raw p = 0.4998, FDR p = 0.5 (not significant)\n",
      "NP_001116224.1: raw p = 0.4527, FDR p = 0.5 (not significant)\n",
      "NP_001116838.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001121167.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001121172.1: raw p = 0.0752, FDR p = 0.5 (not significant)\n",
      "NP_001121173.1: raw p = 0.0748, FDR p = 0.5 (not significant)\n",
      "NP_001121174.1: raw p = 0.075, FDR p = 0.5 (not significant)\n",
      "NP_001121175.1: raw p = 0.0753, FDR p = 0.5 (not significant)\n",
      "NP_001121176.1: raw p = 0.0749, FDR p = 0.5 (not significant)\n",
      "NP_001121177.1: raw p = 0.0749, FDR p = 0.5 (not significant)\n",
      "NP_001121178.1: raw p = 0.0749, FDR p = 0.5 (not significant)\n",
      "NP_001121179.1: raw p = 0.0748, FDR p = 0.5 (not significant)\n",
      "NP_001121363.1: raw p = 0.1729, FDR p = 0.5 (not significant)\n",
      "NP_001121676.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001121677.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001121678.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001127890.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001130000.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001130002.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001136067.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001136068.1: raw p = 0.0416, FDR p = 0.5 (not significant)\n",
      "NP_001136070.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001136071.1: raw p = 0.0181, FDR p = 0.3172 (not significant)\n",
      "NP_001138503.1: raw p = 0.4647, FDR p = 0.5 (not significant)\n",
      "NP_001138743.1: raw p = 0.1686, FDR p = 0.5 (not significant)\n",
      "NP_001138744.1: raw p = 0.1694, FDR p = 0.5 (not significant)\n",
      "NP_001138765.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001138767.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001138792.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001139761.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001153059.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001153060.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001155313.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001157961.1: raw p = 0.3958, FDR p = 0.5 (not significant)\n",
      "NP_001157962.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159392.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159393.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159716.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159717.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159906.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159907.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001159921.1: raw p = 0.0085, FDR p = 0.3094 (not significant)\n",
      "NP_001159923.1: raw p = 0.2222, FDR p = 0.5 (not significant)\n",
      "NP_001159970.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001161066.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001167515.1: raw p = 0.0001, FDR p = 0.006067 (SIGNIFICANT)\n",
      "NP_001167516.1: raw p = 0.0001, FDR p = 0.006067 (SIGNIFICANT)\n",
      "NP_001167517.1: raw p = 0.0001, FDR p = 0.006067 (SIGNIFICANT)\n",
      "NP_001171131.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001177411.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001182220.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001184026.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001184027.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001186668.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001186759.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001186958.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001186959.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001186960.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001186961.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001188504.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001191689.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001193943.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001226.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001230175.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001244900.1: raw p = 0.4042, FDR p = 0.5 (not significant)\n",
      "NP_001244901.1: raw p = 0.3579, FDR p = 0.5 (not significant)\n",
      "NP_001248393.1: raw p = 0.4583, FDR p = 0.5 (not significant)\n",
      "NP_001248759.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001248760.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001258751.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001258752.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001258753.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001258754.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001263419.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001268659.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001269376.1: raw p = 0.4884, FDR p = 0.5 (not significant)\n",
      "NP_001269604.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001271204.1: raw p = 0.2823, FDR p = 0.5 (not significant)\n",
      "NP_001271205.1: raw p = 0.2836, FDR p = 0.5 (not significant)\n",
      "NP_001275536.1: raw p = 0.4745, FDR p = 0.5 (not significant)\n",
      "NP_001275537.1: raw p = 0.4905, FDR p = 0.5 (not significant)\n",
      "NP_001275543.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001275961.1: raw p = 0.4146, FDR p = 0.5 (not significant)\n",
      "NP_001275962.1: raw p = 0.4164, FDR p = 0.5 (not significant)\n",
      "NP_001276093.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001276094.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001278207.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001278208.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001278238.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001313.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001314.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001324.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001325.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001326.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001327.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001624.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001747.2: raw p = 0.4954, FDR p = 0.5 (not significant)\n",
      "NP_001805.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001870.3: raw p = 0.4621, FDR p = 0.5 (not significant)\n",
      "NP_001889.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001890.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001891.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001900.1: raw p = 0.2305, FDR p = 0.5 (not significant)\n",
      "NP_001902.1: raw p = 0.2819, FDR p = 0.5 (not significant)\n",
      "NP_001903.1: raw p = 0.3498, FDR p = 0.5 (not significant)\n",
      "NP_001963.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_001985.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002206.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002207.2: raw p = 0.3895, FDR p = 0.5 (not significant)\n",
      "NP_002209.2: raw p = 0.0002, FDR p = 0.0104 (SIGNIFICANT)\n",
      "NP_002308.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002309.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002413.1: raw p = 0.4262, FDR p = 0.5 (not significant)\n",
      "NP_002414.1: raw p = 0.1829, FDR p = 0.5 (not significant)\n",
      "NP_002415.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002417.2: raw p = 0.1192, FDR p = 0.5 (not significant)\n",
      "NP_002418.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002419.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002572.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002606.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002629.1: raw p = 0.0042, FDR p = 0.1699 (not significant)\n",
      "NP_002630.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002631.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002649.1: raw p = 0.3499, FDR p = 0.5 (not significant)\n",
      "NP_002760.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002761.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002762.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002763.2: raw p = 0.48, FDR p = 0.5 (not significant)\n",
      "NP_002766.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_002965.1: raw p = 0.1254, FDR p = 0.5 (not significant)\n",
      "NP_003055.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003108.2: raw p = 0.0001, FDR p = 0.006067 (SIGNIFICANT)\n",
      "NP_003236.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003245.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003246.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003247.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003540.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003610.2: raw p = 0.0293, FDR p = 0.431 (not significant)\n",
      "NP_003764.3: raw p = 0.4348, FDR p = 0.5 (not significant)\n",
      "NP_003775.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003784.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003803.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003804.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003805.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_003806.3: raw p = 0.4804, FDR p = 0.5 (not significant)\n",
      "NP_003808.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004070.3: raw p = 0.4261, FDR p = 0.5 (not significant)\n",
      "NP_004123.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004146.1: raw p = 0.473, FDR p = 0.5 (not significant)\n",
      "NP_004190.1: raw p = 0.0183, FDR p = 0.3172 (not significant)\n",
      "NP_004236.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004381.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004521.1: raw p = 0.1725, FDR p = 0.5 (not significant)\n",
      "NP_004559.4: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004762.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004985.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_004986.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005015.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005016.1: raw p = 0.4527, FDR p = 0.5 (not significant)\n",
      "NP_005204.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005483.1: raw p = 0.4626, FDR p = 0.5 (not significant)\n",
      "NP_005567.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005568.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005579.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005738.4: raw p = 0.3399, FDR p = 0.5 (not significant)\n",
      "NP_005916.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_005932.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_006191.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_006206.2: raw p = 0.4149, FDR p = 0.5 (not significant)\n",
      "NP_006207.1: raw p = 0.4709, FDR p = 0.5 (not significant)\n",
      "NP_006208.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_006211.2: raw p = 0.0748, FDR p = 0.5 (not significant)\n",
      "NP_006601.2: raw p = 0.2912, FDR p = 0.5 (not significant)\n",
      "NP_006656.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_006681.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_008850.1: raw p = 0.3282, FDR p = 0.5 (not significant)\n",
      "NP_008914.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_008968.4: raw p = 0.1764, FDR p = 0.5 (not significant)\n",
      "NP_008969.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_031369.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_031378.1: raw p = 0.2849, FDR p = 0.5 (not significant)\n",
      "NP_036401.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_036529.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_036596.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_055058.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_055294.1: raw p = 0.1843, FDR p = 0.5 (not significant)\n",
      "NP_055509.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_055679.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_055985.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_056245.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_056933.2: raw p = 0.0135, FDR p = 0.3172 (not significant)\n",
      "NP_057239.4: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_057270.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_060662.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_060703.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_061325.1: raw p = 0.3127, FDR p = 0.5 (not significant)\n",
      "NP_061905.2: raw p = 0.4986, FDR p = 0.5 (not significant)\n",
      "NP_062562.1: raw p = 0.4703, FDR p = 0.5 (not significant)\n",
      "NP_064608.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_064714.2: raw p = 0.4999, FDR p = 0.5 (not significant)\n",
      "NP_067610.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_068367.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_068368.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_068573.2: raw p = 0.3273, FDR p = 0.5 (not significant)\n",
      "NP_068600.4: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_068755.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_068813.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_071356.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_071405.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_071751.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_071913.1: raw p = 0.0129, FDR p = 0.3172 (not significant)\n",
      "NP_079496.1: raw p = 0.4914, FDR p = 0.5 (not significant)\n",
      "NP_085046.5: raw p = 0.4294, FDR p = 0.5 (not significant)\n",
      "NP_109591.1: raw p = 0.1265, FDR p = 0.5 (not significant)\n",
      "NP_112191.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_112217.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_115587.6: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_115992.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_127509.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_149348.2: raw p = 0.0617, FDR p = 0.5 (not significant)\n",
      "NP_149349.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_150377.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_254275.1: raw p = 0.014, FDR p = 0.3172 (not significant)\n",
      "NP_443098.3: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_443187.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_444272.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_536723.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_542177.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_543020.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_570612.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_598000.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_598377.3: raw p = 0.2354, FDR p = 0.5 (not significant)\n",
      "NP_612140.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_620686.1: raw p = 0.4744, FDR p = 0.5 (not significant)\n",
      "NP_624302.1: raw p = 0.2142, FDR p = 0.5 (not significant)\n",
      "NP_631947.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_666023.1: raw p = 0.3581, FDR p = 0.5 (not significant)\n",
      "NP_671724.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_680475.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_694859.1: raw p = 0.0001, FDR p = 0.006067 (SIGNIFICANT)\n",
      "NP_694882.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_695013.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_695014.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_695015.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_695017.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_710159.1: raw p = 0.3323, FDR p = 0.5 (not significant)\n",
      "NP_776249.1: raw p = 0.1786, FDR p = 0.5 (not significant)\n",
      "NP_783866.2: raw p = 0.2827, FDR p = 0.5 (not significant)\n",
      "NP_808807.2: raw p = 0.2791, FDR p = 0.5 (not significant)\n",
      "NP_808808.1: raw p = 0.2788, FDR p = 0.5 (not significant)\n",
      "NP_878907.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_891550.1: raw p = 0.4733, FDR p = 0.5 (not significant)\n",
      "NP_922932.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_940912.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_940998.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_942130.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_945345.2: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_997077.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "NP_997079.1: raw p = 0.4874, FDR p = 0.5 (not significant)\n",
      "NP_997080.1: raw p = 0.4941, FDR p = 0.5 (not significant)\n",
      "XP_005245479.1: raw p = 0.4995, FDR p = 0.5 (not significant)\n",
      "XP_005246699.1: raw p = 0.4707, FDR p = 0.5 (not significant)\n",
      "XP_005246989.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005248097.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005248716.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005249241.1: raw p = 0.4838, FDR p = 0.5 (not significant)\n",
      "XP_005249871.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005250295.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005251773.1: raw p = 0.3575, FDR p = 0.5 (not significant)\n",
      "XP_005256758.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005258332.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005259574.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005260515.1: raw p = 0.3124, FDR p = 0.5 (not significant)\n",
      "XP_005260696.1: raw p = 0.4467, FDR p = 0.5 (not significant)\n",
      "XP_005261654.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005265386.1: raw p = 0.3169, FDR p = 0.5 (not significant)\n",
      "XP_005265581.1: raw p = 0.0621, FDR p = 0.5 (not significant)\n",
      "XP_005265582.1: raw p = 0.0626, FDR p = 0.5 (not significant)\n",
      "XP_005266060.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005266835.1: raw p = 0.0293, FDR p = 0.431 (not significant)\n",
      "XP_005267408.1: raw p = 0.1726, FDR p = 0.5 (not significant)\n",
      "XP_005267790.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005268208.1: raw p = 0.4017, FDR p = 0.5 (not significant)\n",
      "XP_005268209.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005268355.1: raw p = 0.0525, FDR p = 0.5 (not significant)\n",
      "XP_005269337.1: raw p = 0.2783, FDR p = 0.5 (not significant)\n",
      "XP_005270250.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005272173.1: raw p = 0.0174, FDR p = 0.3172 (not significant)\n",
      "XP_005272174.1: raw p = 0.0182, FDR p = 0.3172 (not significant)\n",
      "XP_005272175.1: raw p = 0.0181, FDR p = 0.3172 (not significant)\n",
      "XP_005272176.1: raw p = 0.0177, FDR p = 0.3172 (not significant)\n",
      "XP_005272177.1: raw p = 0.0183, FDR p = 0.3172 (not significant)\n",
      "XP_005272294.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005272296.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005272702.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_005273727.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006713720.1: raw p = 0.4515, FDR p = 0.5 (not significant)\n",
      "XP_006713764.1: raw p = 0.4245, FDR p = 0.5 (not significant)\n",
      "XP_006714791.1: raw p = 0.018, FDR p = 0.3172 (not significant)\n",
      "XP_006714792.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006714793.1: raw p = 0.1003, FDR p = 0.5 (not significant)\n",
      "XP_006715071.1: raw p = 0.1266, FDR p = 0.5 (not significant)\n",
      "XP_006715073.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006717023.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006717400.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006718790.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006718791.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006718792.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006718821.1: raw p = 0.1668, FDR p = 0.5 (not significant)\n",
      "XP_006720222.1: raw p = 0.4996, FDR p = 0.5 (not significant)\n",
      "XP_006722625.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006723893.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006725226.1: raw p = 0.5, FDR p = 0.5 (not significant)\n",
      "XP_006725227.1: raw p = 0.4999, FDR p = 0.5 (not significant)\n",
      "XP_006725460.1: raw p = 0.5, FDR p = 0.5 (not significant)\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import os\n",
    "import re\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# Directory containing your .log files\n",
    "log_dir = 'batecmreg_busted'\n",
    "\n",
    "# Output file\n",
    "output_file = 'batecmreg_bustede1_FDR.txt'\n",
    "\n",
    "# We will store all (protein_name, p_value) in a list for later correction\n",
    "pvals_data = []\n",
    "\n",
    "# 1. Parse each .log file and extract p-values\n",
    "for filename in os.listdir(log_dir):\n",
    "    if filename.endswith('.log'):\n",
    "        filepath = os.path.join(log_dir, filename)\n",
    "        # Extract the protein name from the filename\n",
    "        parts = filename.split('_')\n",
    "        if parts[0].startswith('NP') or parts[0].startswith('XP'):\n",
    "            protein_name = '_'.join(parts[:2])\n",
    "        else:\n",
    "            protein_name = parts[0]\n",
    "\n",
    "        # Read the log file\n",
    "        with open(filepath, 'r') as log_f:\n",
    "            lines = log_f.readlines()\n",
    "            found_pval = False\n",
    "            for line in lines:\n",
    "                if 'Likelihood ratio test for episodic diversifying positive selection' in line:\n",
    "                    # Extract the p-value\n",
    "                    match = re.search(r'p =\\s*([\\d\\.eE+-]+)', line)\n",
    "                    if match:\n",
    "                        p_value_str = match.group(1)\n",
    "                        try:\n",
    "                            p_value = float(p_value_str)\n",
    "                            # Store this (protein, p-value)\n",
    "                            pvals_data.append((protein_name, p_value))\n",
    "                            found_pval = True\n",
    "                        except ValueError:\n",
    "                            print(f\"{protein_name}: Could not convert p-value '{p_value_str}' to float\")\n",
    "                    else:\n",
    "                        print(f\"{protein_name}: p-value not found in line: {line.strip()}\")\n",
    "                    break  # Stop after finding the line\n",
    "            if not found_pval:\n",
    "                print(f\"{protein_name}: 'Likelihood ratio test' line not found or no p-value in {filename}\")\n",
    "\n",
    "# 2. Once we've collected all p-values, apply FDR correction\n",
    "p_values = [x[1] for x in pvals_data]  # just the numeric p-values\n",
    "\n",
    "if len(p_values) == 0:\n",
    "    print(\"No p-values found. Exiting.\")\n",
    "    exit()\n",
    "\n",
    "# Benjamini-Hochberg procedure (FDR)\n",
    "# returns: reject (boolean array), pvals_corrected, alphacSidak, alphacBonf\n",
    "reject_array, pvals_corrected, _, _ = multipletests(p_values, alpha=0.05, method='fdr_bh')\n",
    "\n",
    "# 3. Write results to file (only those that pass FDR, i.e., q < 0.05)\n",
    "with open(output_file, 'w') as out_f:\n",
    "    for i, (protein_name, original_pval) in enumerate(pvals_data):\n",
    "        fdr_pval = pvals_corrected[i]\n",
    "        if reject_array[i]:\n",
    "            # This means fdr_pval < 0.05\n",
    "            out_f.write(protein_name + '\\n')\n",
    "            print(f\"{protein_name}: raw p = {original_pval:.4g}, FDR p = {fdr_pval:.4g} (SIGNIFICANT)\")\n",
    "        else:\n",
    "            print(f\"{protein_name}: raw p = {original_pval:.4g}, FDR p = {fdr_pval:.4g} (not significant)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "EXAMINATION OF FITMG94 and R part"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
